src/lib/core/data-table/data-table.ts

```diff
@@ -0,0 +1,189 @@
+  @ViewChild(CdkHeaderRowPlaceholder) _headerRowPlaceholder: CdkHeaderRowPlaceholder;
+
+  /**
+   * The column definitions provided by the user that contain what the header and cells should
+   * render for each column.
+   */
+  @ContentChildren(CdkColumnDef) _columnDefinitions: QueryList<CdkColumnDef>;
+
+  /** Template used as the header container. */
+  @ContentChild(CdkHeaderRowDef) _headerDefinition: CdkHeaderRowDef;
+
+  /** Set of templates that used as the data row containers. */
+  @ContentChildren(CdkRowDef) _rowDefinitions: QueryList<CdkRowDef>;
+
+  constructor(private _changeDetectorRef: ChangeDetectorRef) {}
+
+  ngOnDestroy() {
+    // TODO(andrewseguin): Disconnect from the data source so
+    // that it can unsubscribe from its streams.
+  }
+
+  ngOnInit() {
+    // TODO(andrewseguin): Setup a listener for scroll events
+    // and emit the calculated view to this.viewChanges
+  }
+
+  ngAfterContentInit() {
+    // TODO(andrewseguin): Throw an error if two columns share the same name
+    this._columnDefinitions.forEach(columnDef => {
+      this._columnDefinitionsByName.set(columnDef.name, columnDef);
+    });
+  }
+
+  ngAfterViewInit() {
+    // TODO(andrewseguin): Re-render the header when the header's columns change.
+    this.renderHeaderRow();
+
+    // TODO(andrewseguin): Re-render rows when their list of columns change.
+    // TODO(andrewseguin): If the data source is not
+    // present after view init, connect it when it is defined.
+    this.dataSource.connectTable(this.viewChanges).subscribe((rowsData: any[]) => {
+      // TODO(andrewseguin): Add a differ that will check if the data has changed,
+      // rather than re-rendering all rows
+      this._rowPlaceholder.viewContainer.clear();
+      rowsData.forEach(rowData => this.insertRow(rowData));
+      this._changeDetectorRef.markForCheck();
+    });
+  }
+
+  /**
+   * Create the embedded view for the header template and place it in the header row view container.
+   */
+  renderHeaderRow() {
+    const cells = this.getHeaderCellTemplatesForRow(this._headerDefinition);
+
+    // TODO(andrewseguin): add some code to enforce that exactly
+    // one CdkCellOutlet was instantiated as a result
+    // of `createEmbeddedView`.
+    this._headerRowPlaceholder.viewContainer
+        .createEmbeddedView(this._headerDefinition.template, {cells});
+    CdkCellOutlet.mostRecentCellOutlet.cells = cells;
+    CdkCellOutlet.mostRecentCellOutlet.context = {};
+  }
+
+  /**
+   * Create the embedded view for the data row template and place it in the correct index location
+   * within the data row view container.
+   */
+  insertRow(rowData: any) {
+    // TODO(andrewseguin): Add when predicates to the row definitions
+    // to find the right template to used based on
+    // the data rather than choosing the first row definition.
+    const row = this._rowDefinitions.first;
+
+    const context = {$implicit: rowData};
```

Comment:
Consider adding some of the other ngFor context stuff, like first, last, even, and odd. I can imagine wanting to make a table with alternating colored rows, where even would be useful.

(Or add a TODO)